// @umarise/anchor/dist/index.js
// src/index.ts
var UmariseCoreError = class extends Error {
  code;
  statusCode;
  retryAfterSeconds;
  constructor(code, message, statusCode, retryAfterSeconds) {
    super(message);
    this.name = "UmariseCoreError";
    this.code = code;
    this.statusCode = statusCode;
    this.retryAfterSeconds = retryAfterSeconds;
  }
};
var UmariseCore = class {
  apiKey;
  baseUrl;
  timeout;
  constructor(config = {}) {
    this.apiKey = config.apiKey;
    this.baseUrl = (config.baseUrl || "https://lppltmdtiypbfzlszhhb.supabase.co/functions/v1").replace(/\/$/, "");
    this.timeout = config.timeout || 3e4;
  }
  // ─── Public Endpoints ───────────────────────────────────────
  /**
   * Check API health.
   * 
   * @returns { status: 'operational', version: 'v1' }
   */
  async health() {
    const res = await this.request("GET", "/v1-core-health");
    return res;
  }
  /**
   * Resolve an origin by ID or hash.
   * Returns the earliest attestation (first-in-time).
   * 
   * @param params - Either { originId } or { hash }
   * @returns Origin record, or null if not found
   */
  async resolve(params) {
    const query = "originId" in params ? `origin_id=${encodeURIComponent(params.originId)}` : `hash=${encodeURIComponent(normalizeHashInput(params.hash))}`;
    try {
      return await this.request("GET", `/v1-core-resolve?${query}`);
    } catch (err) {
      if (err instanceof UmariseCoreError && err.statusCode === 404) return null;
      throw err;
    }
  }
  /**
   * Verify a hash against the registry.
   * Returns the origin record if found, null if no match.
   * 
   * @param hash - SHA-256 hash (with or without 'sha256:' prefix)
   * @returns Origin record with proof status, or null
   */
  async verify(hash) {
    try {
      return await this.request("POST", "/v1-core-verify", {
        hash: normalizeHashInput(hash)
      });
    } catch (err) {
      if (err instanceof UmariseCoreError && err.statusCode === 404) return null;
      throw err;
    }
  }
  /**
   * Download the OpenTimestamps proof for an origin.
   * 
   * @param originId - UUID of the origin
   * @returns Proof result with binary .ots data (or pending/not_found status)
   */
  async proof(originId) {
    const url = `${this.baseUrl}/v1-core-proof?origin_id=${encodeURIComponent(originId)}`;
    const res = await this.fetchWithTimeout(url, { method: "GET", headers: {} });
    if (res.status === 404) {
      return { proof: null, status: "not_found", origin_id: originId };
    }
    if (res.status === 202) {
      return { proof: null, status: "pending", origin_id: originId };
    }
    if (res.status === 200) {
      const bytes = new Uint8Array(await res.arrayBuffer());
      return {
        proof: bytes,
        status: "anchored",
        origin_id: originId,
        bitcoin_block_height: parseInt(res.headers.get("x-bitcoin-block-height") || "0") || void 0,
        anchored_at: res.headers.get("x-anchored-at") || void 0
      };
    }
    await this.handleErrorResponse(res);
    throw new Error("Unreachable");
  }
  // ─── Partner Endpoint (requires API key) ────────────────────
  /**
   * Create an origin attestation.
   * Requires a Partner API key.
   * 
   * @param hash - SHA-256 hash (with or without 'sha256:' prefix)
   * @returns Created origin record
   * @throws UmariseCoreError if no API key is configured
   */
  async attest(hash) {
    if (!this.apiKey) {
      throw new UmariseCoreError(
        "UNAUTHORIZED",
        "API key required for attest(). Pass apiKey in UmariseCore config.",
        401
      );
    }
    return await this.request("POST", "/v1-core-origins", {
      hash: normalizeHashInput(hash)
    }, true);
  }
  // ─── Internal Helpers ───────────────────────────────────────
  async request(method, path, body, authenticated = false) {
    const url = `${this.baseUrl}${path}`;
    const headers = {
      "Content-Type": "application/json"
    };
    if (authenticated && this.apiKey) {
      headers["X-API-Key"] = this.apiKey;
    }
    const res = await this.fetchWithTimeout(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : void 0
    });
    if (res.ok) {
      return res.json();
    }
    await this.handleErrorResponse(res);
  }
  async fetchWithTimeout(url, init) {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);
    try {
      return await fetch(url, { ...init, signal: controller.signal });
    } catch (err) {
      if (err instanceof Error && err.name === "AbortError") {
        throw new UmariseCoreError("TIMEOUT", `Request timed out after ${this.timeout}ms`, 0);
      }
      throw err;
    } finally {
      clearTimeout(timer);
    }
  }
  async handleErrorResponse(res) {
    let errorBody;
    try {
      errorBody = await res.json();
    } catch {
    }
    const err = errorBody?.error;
    throw new UmariseCoreError(
      err?.code || "UNKNOWN_ERROR",
      err?.message || `HTTP ${res.status}`,
      res.status,
      err?.retry_after_seconds
    );
  }
};
function normalizeHashInput(hash) {
  const trimmed = hash.trim().toLowerCase();
  if (trimmed.startsWith("sha256:")) return trimmed;
  if (/^[a-f0-9]{64}$/.test(trimmed)) return `sha256:${trimmed}`;
  return trimmed;
}
async function hashBytes(data) {
  const digest = await globalThis.crypto.subtle.digest("SHA-256", data);
  const hex = Array.from(new Uint8Array(digest)).map((b) => b.toString(16).padStart(2, "0")).join("");
  return `sha256:${hex}`;
}
var index_default = UmariseCore;
export {
  UmariseCore,
  UmariseCoreError,
  index_default as default,
  hashBytes
};
/**
 * Umarise Core SDK — Node.js / TypeScript
 * 
 * Single-file SDK for Umarise Core v1 API.
 * Zero external dependencies. Uses native fetch (Node 18+) or globalThis.fetch.
 * 
 * Usage:
 *   import { UmariseCore } from './umarise-core';
 *   const core = new UmariseCore({ apiKey: 'um_...' });
 *   const origin = await core.attest('sha256:abc123...');
 * 
 * @version 1.0.0
 * @license MIT
 */
